### PR TITLE
Add new world properties (time cycle and weather related features)

### LINKED ISSUE #385: Ability to switch off street lights
This is a request from "zencro":

**Is your feature request related to a problem? Please describe.**
> It's currently impossible to turn off the lights at night time.

**Describe the solution you'd like**
> Perhaps a setAmbientLightingEnabled/toggleAmbientLighting()?

**Describe alternatives you've considered**
> I checked to see if this is possible with shaders and it's not. Disabling `coronastar` makes the lampposts not shine, but it doesn't get rid of the actual light on the ground and also ends up disabling the stars and the sun. Maybe there's a more complex shader solution than that though.
>
> I think a function that makes everything darker would be more useful than something that disables street lights because even with all the street lights removed the world is still way too bright at night time and the only way to get close to proper darkness is one of those super dark weathers. Though this is almost realistic dark:
>
> ```
> crun setWeather(159)
> crun setSkyGradient(0, 0, 0, 0, 0, 0)
> ```

**Additional context**
> It's precisely for roleplay, and mainly for ambience. If a roleplay is set during a blackout, it would be nice to actually not have lights.

From https://bugs.mtasa.com/view.php?id=5543

### LINKED ISSUE #2723: "New" shadows flicker when they are far away from the camera, add an option to change them
### Is your feature request related to a problem? Please describe.

On #1699, TFP worked on a setting that let you lower the shadows in game. Original SA has the shadows a bit too high. This fix works and makes shadows not appear floatty, but this makes them flicker.

There were plans to add a function to change the shadows height, or at least let the client decided whenever they would want the fix or not.

#2018 added this, but without an option to toggle it, since botder argued that this was a general improvement and no one would want it to be disabled. I disagree, because since this change I have been noticing a lot of flickering on far away shadows. I'm 90% sure this didn't happen before, but I could be wrong since this change was added so far ago.

https://user-images.githubusercontent.com/23291573/187042864-c3c0673d-73a6-4b91-88a7-4a3501a00bb1.mp4

We should have an option to enable or disable this, or even better, set the shadow height to whatever we want.

### Describe the solution you'd like

Consider adding the setting proposed on #1699.

### Describe alternatives you've considered

Alternatively add a clientside function in which we can set the shadows height.

### Additional context

This visual error COULD be happening for other reason that #2018 (particulary that broken traffic light), I didn't test older versions and I didn't try to compile MTA with the default shadows. If you have encountered this issue, please let me know.

### Security Policy

- [X] I have read and understood the Security Policy and this issue is not about a cheat or security vulnerability.

### LINKED ISSUE #2778: Make volumetric shadows slighty darker
### Is your feature request related to a problem? Please describe.

Volumetric shadows look ugly IMO. These would look a lot better if were blurred and little darker like blob shadows and dynamic ped shadows. Blurring them may be hard but changing their color is one byte change.

### Describe the solution you'd like

Make volumetric shadows a bit darker, maybe? It would be a _breaking_ (?) change, like, it would be different than it is in single player, so we could discuss that but I think these would seriously look better.

### Describe alternatives you've considered

_No response_

### Additional context

This is how blob shadows look: https://i.imgur.com/IdIy20b.png
This is how it looks now (initial alpha set to 50): https://i.imgur.com/Mu1gEVL.png
This is how it looks with alpha set to 95:
https://i.imgur.com/xpB4Eeq.png
https://i.imgur.com/WqZ61bd.png

This is how it looks with alpha set to 85:
https://i.imgur.com/FKnUPBS.png

### Security Policy

- [X] I have read and understood the Security Policy and this issue is not about a cheat or security vulnerability.

### LINKED ISSUE #2785: Allow custom interval for setWeatherBlended
### Is your feature request related to a problem? Please describe.

Currently if you change minute duration setWeatherBlended is unuseable.

### Describe the solution you'd like

```
setWeather(0)
setWeatherBlended(9, 10000) -- changes blended weather in 10 secs
```

### Describe alternatives you've considered

```
setWeatherBlendTime()
resetWeatherBlendTime()
```

### Additional context

_No response_

### Security Policy

- [X] I have read and understood the Security Policy and this issue is not about a cheat or security vulnerability.

### PR DESCRIPTION
**Pull request by samr46**
This adds `getWorldProperty`, `setWorldProperty`, `resetWorldProperty` to modify remaining time cycle and weather related properties. Unlike `setTimeCycle` from stale PR #276 this implementation is consistent with existing related functions and adds more features.

Imgur album: https://imgur.com/a/jeUxx7L

| Property | Description |
| --- | --- |
| AmbientColor | The color of ambient light on map objects (including custom objects) |
| AmbientObjColor | The color of ambient light on dynamically created elements (peds, vehicles) |
| DirectionalColor | The color of direct light on dynamically created elements (peds, vehicles) |
| Illumination | Multiplier for the directional light (DirectionalColor) |
| SpriteSize | Point lights corona size (looks like only traffic lights are affected) |
| SpriteBrightness | Point lights corona brightness (also affects light on ground for light poles) |
| LightsOnGround | Point lights ground reflection brightness (looks like only traffic lights are affected) |
| PoleShadowStrength | Pole shadows alpha (used if volumetric shadows are disabled) |
| ShadowStrength | Shadows alpha (all shadows) |
| ShadowsOffset | Shadows height<br>Added for convenience to adjust shadows height on demand. Can be used to resolve #2723 |
| BottomCloudsColor | Bottom (normal) clouds color |
| LowCloudsColor | Low (skyline) clouds color (seems to be dependant on game hours) |
| CloudsAlpha | Bottom (normal) clouds alpha |
| | |
| WetRoads | Wet roads weather effect (mostly noticeable during driving)<br>Turns on lights on ground during daytime |
| Foggyness | Adds light fog effect for headlights and turns on lights on ground during daytime<br>Also affects skyline clouds and shadows visibility |
| Fog | Fog weather effect alpha |
| RainFog | Rain fog weather effect alpha (different fog; used in rain weathers) (reset is smooth) |
| WaterFog | Water fog alpha |
| Rainbow | Rainbow alpha |
| Sandstorm | Sandstorm sound volume (reset is smooth) |

Some properties like `AmbientColor` are float by default and adjusted to regular RGB for usability (consistent with timecyc.dat).

<hr>

### Discarded properties:

`ExtraSunnyness`, `CloudCoverage`
> Both properties work with `Foggyness` to determine visiblity of skyline clouds. Might actually be necessary if someone wants to use default skyline clouds and alter this behavior of default weathers. But most servers tend to use skydomes with custom clouds these days. And if needed I'd rather we add separate property that could change skyline clouds alpha.

`CloudsAlpha2`, `CloudsAlpha3`
> Appears to be useless. Changing `CloudsAlpha2` doesn't produce any visible results. `CloudsAlpha3` slightly changes water fog alpha which is redundant.

### Tested and can be added:

`TrafficLightsBrightness` or `TrafficLights`
> Changes traffic lights ground reflection brightness. By default game uses `WetRoads`, `Foggyness`, `Rain level` and game hour properties to determine if light reflections should be visible. So this could act like main override.

<hr>

Fixes #385
> setWorldProperty("SpriteBrightness", 0)

Fixes #2778
> setWorldProperty("ShadowStrength", 500)

Fixes #2785 (kind of)
> This function will allow developers to create custom weathers utilizing all time cycle and weather related properties (along with existing functions). So custom blending, 24h time cycles, anything is possible.

<hr>

Using new parser for this was tricky. But I think it looks good.

It shouldn't be too hard to review. Offsets are used to make memory address validation easier. Test resource with GUI.

Test resource: [worldprops.zip](https://github.com/multitheftauto/mtasa-blue/files/11375576/worldprops.zip)

### CODE REVIEW

**Comment by Fernando-A-Rocha** [comment]
Wow amazing 😄

**Comment by FileEX** [comment]
I really don't understand why no one has looked into this for a year. It's a pity that so few people have write access

File: Client/game_sa/CWeatherSA.cpp
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.cpp:68]: Maybe make it a `#define` so it could be reused later?

File: Client/game_sa/CWeatherSA.cpp
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.cpp:94]: Above

File: Client/game_sa/CWeatherSA.cpp
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.cpp:118]: Above

File: Client/game_sa/CWeatherSA.cpp
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.cpp:142]: Above

File: Client/game_sa/CWeatherSA.cpp
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.cpp:164]: Above

File: Client/game_sa/CWeatherSA.cpp
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.cpp:188]: Above

File: Client/game_sa/CWeatherSA.cpp
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.cpp:212]: Above

File: Client/game_sa/CWeatherSA.h
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.h]: Please add `const` specifier for getters

File: Client/game_sa/CWeatherSA.h
Comment (TracerDS) [inline Client/game_sa/CWeatherSA.h:61]: `std::uint8_t` instead of `unsigned char`

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (TracerDS) [inline Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp:2087]: `std::int16_t` instead of `int16`

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (TracerDS) [inline Client/multiplayer_sa/CMultiplayerSA.cpp:354]: `std::int16_t` instead of `int16`

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (TracerDS) [inline Client/multiplayer_sa/CMultiplayerSA.cpp:2268]: `#define` for mem addresses maybe?

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (TracerDS) [inline Client/multiplayer_sa/CMultiplayerSA.cpp:2291]: Above

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (TracerDS) [inline Client/multiplayer_sa/CMultiplayerSA.cpp:2314]: Above

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (TracerDS) [inline Client/multiplayer_sa/CMultiplayerSA.cpp:2357]: Above

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (TracerDS) [inline Client/multiplayer_sa/CMultiplayerSA.cpp:2380]: Above

File: Client/multiplayer_sa/CMultiplayerSA.h
Comment (TracerDS) [inline Client/multiplayer_sa/CMultiplayerSA.h]: `const` qualifier for getters, `std::int16_t` instead of `int16`

**Review by TracerDS — CHANGES_REQUESTED** [review]
Good job

**Comment by samr46** [comment]
@TracerDS
I didn't put memory addresses into definitions (and don't want to) to keep consistency within the file and with current implementation of similar functions. I don't see any reason to use #define for memory addresses in this case because it will make code messy and review harder.

I will make changes to getters and data types later today. Thank you.
Originally I didn't use `const` and `override` to make methods as similar as possible to other methods within the file (again for consistency). It actually looks weird right now after we started to use it in newer PRs without proper refactor of old code.

**Comment by TracerDS** [comment]
> @TracerDS I didn't put memory addresses into definitions (and don't want to) to keep consistency within the file and with current implementation of similar functions. I don't see any reason to use #define for memory addresses in this case because it will make code messy and review harder.

Magic values are already a hindrance in mta source code. Having them as a definition makes it easier to understand

> I will make changes to getters and data types later today. Thank you. Originally I didn't use `const` and `override` to make methods as similar as possible to other methods within the file (again for consistency). It actually looks weird right now after we started to use it in newer PRs without proper refactor of old code.

👌

**Comment by samr46** [comment]
> Having them as a definition makes it easier to understand

I actually agree with this and it's useful for reusing address within getter and setter. But it's harder to copy memory address during review / code inspection. And the last one is important since we have PRs without review for years due limited interest / manpower.

**Comment by TracerDS** [comment]
> But it's harder to copy memory address during review / code inspection.

How so?

**Comment by samr46** [comment]
> How so?

Double click and Ctrl + C is faster and more intuitive than hovering and clicking Copy or switching directly to #define value. The only benefit is that it can be easier to understand in some cases and less risk of using wrong address in multiple places. In this case all addresses are related to method names so we don't really have any problem with understanding what it will change.

**Comment by TracerDS** [comment]
> Double click and Ctrl + C is faster and more intuitive than hovering and clicking Copy or switching directly to #define value. The only benefit is that it can be easier to understand in some cases and less risk of using wrong address in multiple places. In this case all addresses are related to method names so we don't really have any problem with understanding what it will change.

Thats just laziness that doesnt benefit anyone. Memory addresses dont change, definitions do.

**Comment by samr46** [comment]
> Memory addresses dont change, definitions do.

In our case nothing will really change. This is why it's better to keep it consistent imo.

File: Client/game_sa/CWeatherSA.h
Comment (samr46) [reply Client/game_sa/CWeatherSA.h:61]: By the way, I'm not the author of these changes. It was merged from master in a1db8f4 (it looks like some private changes, I don't really know where it originated in and who is the author). It's not even used in public files as far as I know.

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (samr46) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: Technically `std::int16_t` and `int16` are the same (short is signed by default) 🤔
`int16` is more clean and consistent with reversed code

Comment (TracerDS) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: > Technically `std::int16_t` and `int16` are the same (short is signed by default) 🤔 `int16` is more clean and consistent with reversed code

No. They are **not** the same. `std::int16_t` is crossplatform while `int16` is not. Please use `std` variant instead

File: Client/game_sa/CWeatherSA.h
Comment (TracerDS) [reply Client/game_sa/CWeatherSA.h:61]: > By the way, I'm not the author of these changes. It was merged from master in [a1db8f4](https://github.com/multitheftauto/mtasa-blue/commit/a1db8f45913c7bc4b4f7c76fcd0163c9495961f1) (it looks like some private changes, I don't really know where it originated in and who is the author). It's not even used in public files as far as I know.

But a refactor in this case would be better

**Comment by TracerDS** [comment]
> In our case nothing will really change. This is why it's better to keep it consistent imo.

Thats not consistency. Going with harder to understand codeblocks just because is not a good approach.
Having it defined as a preprocessor variable costs us nothing but benefits exceed the cons.

Magic values are **NEVER** the way to go

**Review by FileEX — APPROVED** [review]
LGTM. Debating whether to use a definition for addresses should not be part of this PR. The same applies to discussions about whether to use std library types (recommended for C++) or types from the global namespace (typical for C and available in C++ as well). Such debates should not be part of this PR, as they only lengthen its review. This is not a code refactoring PR. We all know that the MTA code is long spaghetti, but separate PRs are used for refactoring (by the way, they have been waiting for further steps for a long time). Imo code review concerns how the code was written in terms of conventions and consistency with the project, and in this case, consistency was maintained. Mixing it up so that half the code uses definitions and the other half uses direct addresses will only create even more spaghetti.

**Comment by TracerDS** [comment]
Its better to not create problems that would add up in the future. That way there would be less code refactoring PRs.
Better to make changes earlier while there is still chance for merge

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (samr46) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: I'd use it for shared or server-side code. But this code is strictly client-side and not cross-platform. It uses the same definition on Windows.

To be honest with you, I don't agree with almost all code suggestions from you. It's not necessary at all and breaks consistency in modified files. I decided to make some changes only because I'm polite and this useful PR was ignored for a year.
For example I don't want to use #define for memory addresses in this particular PR. I strongly oppose not consistent code in files because it makes it harder to work with in the long run. If this is actually required to merge this PR then I'd like to get a comment from Member / someone who is interested in merging this PR.

Comment (TracerDS) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: So, you'd rather stick to unknown addresses without documenting them...?
Its alright. Your PR, your call, just trying to make the source code more reliable and less messy

**Comment by ghost** [comment]
> Its better to not create problems that would add up in the future. That way there would be less code refactoring PRs. Better to make changes earlier while there is still chance for merge

I agree with you on this, the only problem with it is that these are not clarified in the community, they are not stipulated in the coding guidelines.

If these were all clarified and prescribed in a format readable by anyone, there would basically be no problem with it - it would be really useful in the case of a new implementation, preventing further refactoring. 

I personally think that to avoid future problems it would be worthwhile to use std namespace types, but it takes a lot of people's opinions to decide that, and this PR is not the place for that.

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (FileEX) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: Maybe let's make a small compromise and just add comments above the addresses for now?

Comment (TracerDS) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: Its alright. If it would bother other people then a refactoring PR will be posted.

**Comment by FileEX** [comment]
To start with, the coding guidelines haven't been updated for several years. No official documentation is currently up to date regarding the development of the MTA. There is no one who would take care of it and develop it at the appropriate level. Arguing now on every PR will not solve this problem. We should seriously consider developing coding guidelines and establishing some clear community standards.

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (samr46) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: It's documented in reversed code. And this is actually the only place where it's useful.

There is a reason why other similar functions almost don't use definitions for addresses. For example you don't #define addresses when you NOP the same variable in different places because it makes code messy and hard to work with.
It makes sense to #define the address for overwritten variable. But you can also just put a new line there and keep this overwritten variable separated. And this is actually how it's used in all similar functions right now because it's easy to read and compare with reversed code.

**Comment by samr46** [comment]
> To start with, the coding guidelines haven't been updated for several years. No official documentation is currently up to date regarding the development of the MTA. There is no one who would take care of it and develop it at the appropriate level. Arguing now on every PR will not solve this problem. We should seriously consider developing coding guidelines and establishing some clear community standards.

It's true. This is why in all my PRs I prefer to make code consistent with existing codebase even when it's not really good in some cases. Without proper refactor many things look out of place.

File: Client/multiplayer_sa/CMultiplayerSA.cpp
Comment (TracerDS) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: By reversed code you mean...?

Comment (samr46) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: > By reversed code you mean...?

I'm obviously talking about private `.idb` files of `gta_sa.exe` used for creating features like this.

Comment (TracerDS) [reply Client/multiplayer_sa/CMultiplayerSA.cpp:354]: > > By reversed code you mean...?
> 
> I'm obviously talking about private `.idb` files of `gta_sa.exe` used for creating features like this.

Thought you were talking about gta-reversed for a moment there. Its okay 👍

**Review by ghost — APPROVED** [review]
LGTM 🎉

**Review by TracerDS — APPROVED** [review]
lgtm

**Comment by samr46** [comment]
If someone is interested in merging this for testing in nightly, I will add information to wiki within several hours.

**Comment by TracerDS** [comment]
Resolve conversations, please 😃

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (Lpsd) [inline Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp]: Is this an entirely new Lua function/definition? If so, use `ArgumentParser`, not `ArgumentParserWarn`

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (Lpsd) [inline Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp]: Same here (`ArgumentParser`)

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (Lpsd) [inline Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp:127]: Same here (`ArgumentParser`)

**Review by Lpsd — CHANGES_REQUESTED** [review]

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (samr46) [reply Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp]: This is new Lua function. However, if I use `ArgumentParser` then it will throw an error on incorrect / missing property name, wouldn't it?
It's not a problem for me personally but other similar functions throw warnings right now. Also this function is likely to be used in a loop (may cause unwanted loop breaks if we add new properties in the future and scripter forget to add client version checks)

Comment (Lpsd) [reply Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp]: All new functions should use `ArgumentParser` and are hard error on **usage** failure. Even when modifying existing functions, they should be switched to use `ArgumentParserWarn` if possible.

`ArgumentParserWarn` is just for compatibility so upgraded definitions produce a warning instead of error.

Currently you don't actually `throw` in the function, so you'll want to introduce a `default` case to the switch statement and throw in there, ie:

```cpp
default:
    throw std::invalid_argument{"Property not supported"};
```

You can also remove the `return false;` on the last line of the method, as it's not necessary by introducing this.

We shouldn't worry about user scripts failing because they used it wrong, they should worry about using it correctly ;)

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (TracerDS) [inline Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp:2075]: can you do `return { red * 255, green * 255, blue * 255 };`?

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (samr46) [reply Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp]: It seems that default error message produced by ArgumentParser is more than enough to understand what is wrong in this case. So no need to throw custom error message. Thank you for this information regarding new argument parser.

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (TracerDS) [inline Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp:2096]: use static_cast instead, pleasee

**Review by TracerDS — CHANGES_REQUESTED** [review]

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (samr46) [reply Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp:2075]: Doesn't appear to work with `CLuaMultiReturn` as `std::variant`.

**Comment by Dutchman101** [comment]
Regarding the coding guidelines, there's been some discussion on dev discord, and contributors took a draft initiative: [https://wiki.multitheftauto.com/wiki/New_coding_guidelines](https://wiki.multitheftauto.com/wiki/New_coding_guidelines)

This PR has been reviewed a bunch of times by good people, merging. Thanks to @samr46 for his work to get this highly demanded set of features into MTA.

File: Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp
Comment (Dutchman101) [reply Client/mods/deathmatch/logic/luadefs/CLuaWorldDefs.cpp:2096]: If this is validated, it can be done by addendum.

**Pull request merged**
